command/agent/template/template.go

```diff
@@ -0,0 +1,186 @@
+// Package template is responsible for rendering user supplied templates to
+// disk. The Server type managing the lifecycle of an internal Consul Template
+// Runner
+package template
+
+import (
+	"context"
+	"strings"
+
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/command/agent/config"
+)
+
+// ServerConfig is a config struct for setting up the basic parts of the
+// Server
+type ServerConfig struct {
+	Logger hclog.Logger
+	// Client        *api.Client
+	VaultConf     *config.Vault
+	ExitAfterAuth bool
+
+	Namespace string
+}
+
+// Server manages the Consul Template Runner which renders templates
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// unblockCh is used to block until a template is rendered
+	unblockCh chan struct{}
+
+	// Templates holds the parsed Consul Templates
+	Templates []*ctconfig.TemplateConfig
+
+	DoneCh        chan struct{}
+	logger        hclog.Logger
+	exitAfterAuth bool
+}
+
+// NewServer returns a new configured server
+func NewServer(conf *ServerConfig) (*Server, chan struct{}) {
+	unblock := make(chan struct{})
+	ts := Server{
+		DoneCh:        make(chan struct{}),
+		logger:        conf.Logger,
+		unblockCh:     unblock,
+		config:        conf,
+		exitAfterAuth: conf.ExitAfterAuth,
+	}
+	return &ts, unblock
+}
+
+// Run kicks off the internal Consul Template runner, and listens for changes to
+// the token from the AuthHandler. If Done() is called on the context, shut down
+// the Runner and return
+func (ts *Server) Run(ctx context.Context, incoming chan string, templates []*ctconfig.TemplateConfig) {
+	latestToken := new(string)
+	ts.logger.Info("starting sink server")
+	defer func() {
+		ts.logger.Info("template server stopped")
+		close(ts.DoneCh)
+	}()
+
+	if incoming == nil {
+		panic("incoming channel is nil")
+	}
+
+	// If there are no templates, close the unblockCh
+	if len(templates) == 0 {
+		// nothing to do
+		close(ts.unblockCh)
+		return
+	}
+
+	// construct a consul template vault config based the agents vault
+	// configuration
+	var runnerConfig *ctconfig.Config
+	if runnerConfig = newRunnerConfig(ts.config, templates); runnerConfig == nil {
+		ts.logger.Info("template server failed to generate runner config")
+		return
+	}
+
+	var err error
+	ts.runner, err = manager.NewRunner(runnerConfig, false)
+	if err != nil {
+		ts.logger.Info("template server failed to create")
+		return
```

Comment:
Should `close(ts.unblockCh)` be called here? Or deferred at some point earlier on? Also, should there be a panic? Not sure how returning in various ways will effect the behavior of the outer application. Should the logged issue be a warning or error?

Reply:
Good catch, thanks. I mistakenly associated that with the `defer` above that closes the `DoneCh`. I added explicit `close()` calls as well as chanaged the `logger.Info` to `logger.Error` calls, you're right they should not be `Info` 😄 .

Patched in 14daed773

Reply:
I'm going to resolve this for now. I _believe_ I've closed `unblockCh` where it should be, correctly. I'll be able to confirm when I wrap up and push tests